src/javascript/tests/lists.rs

```diff
@@ -82,3 +82,49 @@ const b = [1, [2, [3, []]]];
 "#
     );
 }
+
+#[test]
+fn list_destructuring() {
+    assert_js!(
+        r#"
+fn go(x, y) {
+    let [] = x
+    let [a] = x
+    let [1, 2] = x
+    let [_, #(3, b)] = y
+    let [head, ..tail] = y
+}
+"#,
+        r#""use strict";
+
+function go(x, y) {
+  let gleam$tmp = x;
+  if (!(gleam$tmp?.length === 0)) throw new Error("Bad match");
+  
+  let gleam$tmp = x;
+  if (!(gleam$tmp?.[1]?.length === 0)) throw new Error("Bad match");
+  let a = gleam$tmp[0];
+  
+  let gleam$tmp = x;
+  if (!(
+    gleam$tmp?.[1]?.[1]?.length === 0 &&
+    gleam$tmp[0] === 1 &&
+    gleam$tmp[1][0] === 2
+  )) throw new Error("Bad match");
+  
+  let gleam$tmp = y;
+  if (!(
+    gleam$tmp?.[1]?.[1]?.length === 0 &&
+    gleam$tmp[1][0][0] === 3
+  )) throw new Error("Bad match");
+  let b = gleam$tmp[1][0][1];
+  
+  let gleam$tmp = y;
+  if (!(gleam$tmp?.[1]?.length !== undefined)) throw new Error("Bad match");
```

Comment:
Can do the same here too.

```suggestion
  if (!(gleam$tmp)) throw new Error("Bad match");
```

Reply:
No unfortunately not, the possible return values we have are `0`, `2` and `undefined`. 0 is also falsey but not a case we want to match on. similar to above where we want to match on 0 but not undefined.

Reply:
It can trip into being an undefined that's passed down through the `?` if the list is shorter than needed for the match.

if you list is `[]` then `list?.[1]` is undefined and `undefined?.length` is again undefined.